File: algorithms/hard/longest_increasing_path_in_a_matrix.ex
defmodule Solution do
  @spec longest_increasing_path(matrix :: [[integer]]) :: integer
  def longest_increasing_path(matrix) do
    r = Enum.count(matrix)
    c = Enum.count(hd(matrix))

    {g, dp} =
      matrix
      |> Enum.with_index()
      |> Enum.reduce({%{}, %{}}, fn {row, i}, acc ->
        row
        |> Enum.with_index()
        |> Enum.reduce(acc, fn {n, j}, {g, dp} ->
          {Map.put(g, {i, j}, n), Map.put(dp, {i, j}, 0)}
        end)
      end)

    for i <- 0..(r - 1), j <- 0..(c - 1), reduce: {0, dp} do
      {ans, dp} ->
        dp = dfs(g, dp, i, j)
        ans = max(ans, dp[{i, j}])
        {ans, dp}
    end
    |> elem(0)
  end

  def dfs(g, dp, x, y) do
    if dp[{x, y}] != 0 do
      dp
    else
      for {dx, dy} <- [{-1, 0}, {0, 1}, {1, 0}, {0, -1}], reduce: dp do
        dp ->
          nx = x + dx
          ny = y + dy

          if Map.get(g, {nx, ny}, -1) > g[{x, y}] do
            dp = dfs(g, dp, nx, ny)
            Map.put(dp, {x, y}, max(dp[{x, y}], dp[{nx, ny}]))
          else
            dp
          end
      end
      |> Map.update!({x, y}, &(&1 + 1))
    end
  end
end

File: algorithms/hard/longest_increasing_path_in_a_matrix.rs
fn dfs(mat: &Vec<Vec<i32>>, dp: &mut Vec<Vec<i32>>, x: usize, y: usize) -> i32 {
    if dp[x][y] != 0 {
        return dp[x][y];
    }
    for (dx, dy) in &[(-1, 0), (0, 1), (1, 0), (0, -1)] {
        let nx = (x as i32 + dx) as usize;
        let ny = (y as i32 + dy) as usize;
        if nx >= mat.len() || ny >= mat[0].len() {
            continue;
        }
        if mat[nx][ny] > mat[x][y] {
            dp[x][y] = dp[x][y].max(dfs(mat, dp, nx, ny));
        }        
    }
    dp[x][y] += 1;
    dp[x][y]
}

impl Solution {
    pub fn longest_increasing_path(matrix: Vec<Vec<i32>>) -> i32 {
        let mut dp = vec![vec![0; matrix[0].len()]; matrix.len()];
        let mut ans = 0;
        for i in 0..matrix.len() {
            for j in 0..matrix[0].len() {
                ans = ans.max(dfs(&matrix, &mut dp, i, j))
            }
        }
        ans
    }
}
